### DISCUSSION TITLE
Connect to remote PC

### DISCUSSION THREAD

**Discussion by MaximilianFranz** [Dlubal-Software/RFEM_Python_Client#295] [Q&A]
Hey there,

thanks for the great work on the APIs! 

In the wiki there's [this section on connection to a remote PC.](https://github.com/Dlubal-Software/RFEM_Python_Client/wiki/How-to-connect-to-remote-PC). 
So far so good, but how do I tell my python-HLF library to connect to that IP and port, instead of trying on localhost? 

Do I need to edit the `initModel.py` file manually? (Doesn't seem good practice to change library code)
![image](https://github.com/Dlubal-Software/RFEM_Python_Client/assets/23345908/5812b73f-9ff9-46ed-98dd-1bf546e37b1b)

**Problems Connecting** 
The check supplied in the wiki page for the conenction to the port works just fine.  
I've deactivated the firewall momentarily on the host PC and tried connecting, but it fails somewhere in the `HttpAuthenticated` transport, trying to connect to the `/wsdl` route with a `[Errno 54] Connection reset by peer` 

I made sure RFEM 6 + webservices is running and no dialogs are open on target PC. 

Any hints what could be blocking the connection? 
Or any hint where or if I can see the logs of the Webservices server running on the target machine to see if the connection arrives and is dropped for a specific reason? 

Thanks already for the answer and kind regards, 
Max

**Comment by MaximilianFranz** [comment]
I'm on a Trial Version license, may that be a problem? Because from all I can judge, the connection arrives at the Webservices server, but is closed immediatly.

**Reply by MaximilianFranz** [reply]
Tried with a license that has a Webservices Subscription as well, fails with the same error 54.

**Comment by dogukankaratas** [comment]
Hi,
My suggestion would be uninstalling RFEM package using `pip uninstall RFEM` and use the direct folder root. And please edit the URL section manually. We didn't want to add a feature that asks the URL for every connection.

Also the logs can be found in this path (if it is Windows): c:\Users\xx\AppData\Local\Dlubal\RFEM6_6.02\logs\

Feel free to reach out for further questions.

**Reply by MaximilianFranz** [reply]
With "use the direct folder root" you mean to use the files of repository directly instead of pip installing the `RFEM` python client? 
Will try this and see the logs.

**Reply by MaximilianFranz** [reply]
Logs say: 

```
2023/10/17 11:22:59 556906	JOURNAL [DApplication.cpp->DApplication::notifyImpl:410]	Focus Out: object_name: useLicenseComboBox, value: '01'
2023/10/17 11:23:00 129575	JOURNAL [DApplication.cpp->DApplication::notifyImpl:410]	Mouse Left Click @ [713, 168]: ' {class_name: QWidget}, Parent widget: {class_name: mdi::DMdiArea}, Parent window: {object_name: DMainWindow, title: Dlubal RFEM Testversion | 6.04.0002, class_name: DMainWindow} 
2023/10/17 11:23:11 779633	JOURNAL [DApplication.cpp->DApplication::notifyImpl:410]	Mouse Left Click @ [413, 248]: ' {class_name: QWidget}, Parent widget: {class_name: mdi::DMdiArea}, Parent window: {object_name: DMainWindow, title: Dlubal RFEM Testversion | 6.04.0002, class_name: DMainWindow} 
2023/10/17 11:23:19 518374	SOAP [AbstractSoapServer.cpp->webservices::AbstractSoapServer::log:59]	Port 8081: Incoming connection from ::ffff:XXXXXXXX  is denied. Purchase Web Service Enterprise license to enable it.
2023/10/17 11:23:19 518617	SOAP [AbstractSoapServer.cpp->webservices::AbstractSoapServer::log:59]	Port 8081: Connection rejected
2023/10/17 11:23:19 623594	SOAP [AbstractSoapServer.cpp->webservices::AbstractSoapServer::log:59]	Port 8081: Incoming connection from ::ffff:XXX.XXX.XXX.XXX is denied. Purchase Web Service Enterprise license to enable it.
2023/10/17 11:23:19 623886	SOAP [AbstractSoapServer.cpp->webservices::AbstractSoapServer::log:59]	Port 8081: Connection rejected
2023/10/17 11:23:22 322584	JOURNAL [DApplication.cpp->DApplication::notifyImpl:410]	Mouse Left Click @ [546, 410]: ' {class_name: QWidget}, Parent widget: {class_name: mdi::DMdiArea}, Parent window: {object_name: DMainWindow, title: Dlubal RFEM Testversion | 6.04.0002, class_name: DMainWindow} 
```

So it is a license thing. 
Better user-facing output would be great here. 
Will check again, because we have a Dlubal Webservices API license, but it doesn't seem to show up in my license manager.

**Reply by MaximilianFranz** [reply]
So me again, 

I've ensured I'm on a license on the host PC that has Webservices enabled: 
![image](https://github.com/Dlubal-Software/RFEM_Python_Client/assets/23345908/bd08b897-cba6-4e3e-bd41-21f07c3bc89f)

But still get this log: 
```
2023/10/17 13:41:00 169792	SOAP [AbstractSoapServer.cpp->webservices::AbstractSoapServer::log:59]	Port 8081: Incoming connection from ::ffff:95.90.241.112 is denied. Purchase Web Service Enterprise license to enable it.
```

What's this **Web Service Enterprise**? I can't find anything about it in the webshop or anywhere really. 
Is there something else I might be forgetting about the licensing? 

Thanks already, 
Max

**Reply by MaximilianFranz** [reply]
Hacky workaround I can imagine is writing a small proxy that takes the external requests and forwards them (then coming from localhost) to the webservices server of RFEM in case the feature for external connection is just not enabled really. 
But in terms of robustness I would really like to avoid that.

**Reply by MaximilianFranz** [reply]
As a note: Regarding to the logs, nothing has changed after switching from a trial license to a full license (an restarting RFEM). So mabye the webservices server doesn't recognize the license properly?

**Reply by MaximilianFranz** [reply]
To keep everyone updated: Dlubal is looking into this and providing a fix, which is supposed to be done soon.

**Comment by OndraMichal** [comment]
Hi @MaximilianFranz,
does this issue still persist with the latest updates?

**Reply by MaximilianFranz** [reply]
I will need to check - give me a few days

**Reply by MaximilianFranz** [reply]
Hi @OndraMichal  - Just tested it again. The problem still persists with `1.19.5` of the rfem client. 

Now the logs just read. The connection obviously works. The web-services license is active. 

```
2024/05/16 13:46:10 244135	SOAP [AbstractSoapServer.cpp->webservices::AbstractSoapServer::log:72]	Port 8081: Connection rejected
2024/05/16 13:46:10 298746	SOAP [AbstractSoapServer.cpp->webservices::AbstractSoapServer::log:72]	Port 8081: Connection rejected
``` 

Don't know what else I could change. Any idea? 

Also note that for anyone trying to setup the remote access you should inform them in more detail how to get the connection to the local PC at all. I.e. opening the ports on the router / exposing the local PC to the internet / ...

**Reply by MaximilianFranz** [reply]
Any news here as to what could be the reason for the error logs? @OndraMichal or @dogukankaratas ? :) Thanks tons

**Comment by OndraMichal** [comment]
Hi @MaximilianFranz,
I can't imagine giving more information than what is on the [wiki page](https://github.com/Dlubal-Software/RFEM_Python_Client/wiki/How-to-connect-to-remote-PC) mentioned above. Opening ports on a router or exposing a local PC to the Internet is very individual and should mostly be referred to the IT department. You should do the same so that you don't expose yourself to unnecessary risks.

**Comment by MaximilianFranz** [comment]
Valid point.
 However, I am my IT department. And I can imagine not all of your
customers with 5-20 engineers have a IT administrator that would know right
off the bat that port forwarding is required.

At least some pointers don’t hurt. I don’t expect the setup to be done on
your wiki page.

Thanks :)

Looking forward to your answer on what the underlying issue could be.

Best,
